subsys/net/lib/nrf_cloud/coap/src/nrf_cloud_coap_transport.c

```diff
@@ -0,0 +1,540 @@
+#if defined(CONFIG_POSIX_API)
+#include <zephyr/posix/arpa/inet.h>
+#include <zephyr/posix/netdb.h>
+#include <zephyr/posix/unistd.h>
+#include <zephyr/posix/poll.h>
+#include <zephyr/posix/sys/socket.h>
+#else
+#include <zephyr/net/socket.h>
+#endif
+#include <modem/lte_lc.h>
+#include <zephyr/random/rand32.h>
+#include <nrf_socket.h>
+#include <nrf_modem_at.h>
+#include <date_time.h>
+#include <net/nrf_cloud.h>
+#include <net/nrf_cloud_coap.h>
+#include <cJSON.h>
+#include <version.h>
+#include "nrf_cloud_codec_internal.h"
+#include "dtls.h"
+#include "coap_codec.h"
+#include "nrf_cloud_coap_transport.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(nrf_cloud_coap_transport, CONFIG_NRF_CLOUD_COAP_LOG_LEVEL);
+
+/** @TODO: figure out whether to make this a Kconfig value or place in a header */
+#define CDDL_VERSION "1"
+#define MAX_COAP_PATH 256
+#define MAX_RETRIES 10
+#define JWT_BUF_SZ 700
+#define VER_STRING_FMT "mver=%s&cver=%s&dver=%s"
+#define VER_STRING_FMT2 "cver=" CDDL_VERSION "&dver=" BUILD_VERSION_STR
+#define BUILD_VERSION_STR STRINGIFY(BUILD_VERSION)
+
+static struct sockaddr_storage server;
+
+static int sock = -1;
+static bool authenticated;
+static bool cid_saved;
+
+static struct connection_info
+{
+	uint8_t s4_addr[4];
+	uint8_t d4_addr[4];
+} connection_info;
+
+static struct coap_client coap_client;
+static int nrf_cloud_coap_authenticate(void);
+
+#if defined(CONFIG_NRF_CLOUD_COAP_LOG_LEVEL_DBG)
+static const char *const coap_method_str[] = {
+	NULL,		/* 0 */
+	"GET",		/* COAP_METHOD_GET = 1 */
+	"POST",		/* COAP_METHOD_POST = 2 */
+	"PUT",		/* COAP_METHOD_PUT = 3 */
+	"DELETE",	/* COAP_METHOD_DELETE = 4 */
+	"FETCH",	/* COAP_METHOD_FETCH = 5 */
+	"PATCH",	/* COAP_METHOD_PATCH = 6 */
+	"IPATCH",	/* COAP_METHOD_IPATCH = 7 */
+	NULL
+};
+
+#define METHOD_NAME(method) \
+	(((method >= COAP_METHOD_GET) && (method <= COAP_METHOD_IPATCH)) ?\
+	 coap_method_str[method] : "N/A")
+
+struct format {
+	enum coap_content_format fmt;
+	const char *name;
+};
+
+const struct format formats[] = {
+	{COAP_CONTENT_FORMAT_TEXT_PLAIN, "PLAIN TEXT"},
+	{COAP_CONTENT_FORMAT_APP_LINK_FORMAT, "APP LINK"},
+	{COAP_CONTENT_FORMAT_APP_XML, "XML"},
+	{COAP_CONTENT_FORMAT_APP_OCTET_STREAM, "OCTET STRM"},
+	{COAP_CONTENT_FORMAT_APP_EXI, "EXI"},
+	{COAP_CONTENT_FORMAT_APP_JSON, "JSON"},
+	{COAP_CONTENT_FORMAT_APP_JSON_PATCH_JSON, "JSON PATCH"},
+	{COAP_CONTENT_FORMAT_APP_MERGE_PATCH_JSON, "JSON MRG PATCH"},
+	{COAP_CONTENT_FORMAT_APP_CBOR, "CBOR"}
+};
+
+static const char *fmt_name(enum coap_content_format fmt)
+{
+	for (int i = 0; i < ARRAY_SIZE(formats); i++) {
+		if (formats[i].fmt == fmt) {
+			return formats[i].name;
+		}
+	}
+	return "N/A";
+}
+#endif /* CONFIG_NRF_CLOUD_COAP_LOG_LEVEL_DBG */
+
+bool nrf_cloud_coap_is_connected(void)
+{
+	if (!authenticated) {
+		LOG_DBG("Not connected and authenticated");
+	}
+	return authenticated;
+}
+
+/**@brief Resolves the configured hostname. */
+static int server_resolve(void)
```

Comment:
Would it be possible to instead configure the DTLS socket with a target hostname and let this resolve process be something Zephyr or the DTLS implementation handles? I thought this is how MQTT handles things, but maybe I am misremembering

Reply:
This is how the sdk-nrf/samples/cellular/coap_client works, which was were some of this originated.

The name lookup for nRF Cloud MQTT is done inside nrf_cloud_transport.c in nct_connect().  For nRF Cloud REST, it is done inside the sdk-nrf/subsys/net/lib/rest_client/src/rest_client.c, in the rest_client_sckt_connect() function.

It is not done in Zephyr for the user.